lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Icarus.hs

```diff
@@ -151,6 +172,84 @@ generateKeyFromSeed
     -> IcarusKey 'RootK XPrv
 generateKeyFromSeed = unsafeGenerateKeyFromSeed
 
+-- | Hardware Ledger devices generates keys from mnemonic using a different
+-- approach (different from the rest of Cardano).
+generateKeyFromHardwareLedger
+    :: Mnemonic 24
+        -- ^ The actual seed
+    -> Passphrase "encryption"
+        -- ^ Master encryption passphrase
+    -> IcarusKey 'RootK XPrv
+generateKeyFromHardwareLedger mnemonic (Passphrase pwd) = unsafeFromRight $ do
+    let seed = pbkdf2HmacSha512
+            $ T.encodeUtf8
+            $ T.intercalate " "
+            $ mnemonicToText mnemonic
+
+    let cc = hmacSha256 (BS.pack [1] <> seed)
+    let (kL, kR) = first normalizeEd25519 $ hashRepeatedly seed
+    pA <- ed25519ScalarMult kL
+
+    prv <- left show $ xprv $ kL <> kR <> pA <> cc
+    pure $ IcarusKey (xPrvChangePass (mempty :: ByteString) pwd prv)
+  where
+    unsafeFromRight :: (HasCallStack, Show e) => Either e a -> a
+    unsafeFromRight = either (error . show) id
+
+    hashRepeatedly :: ByteString -> (ByteString, ByteString)
+    hashRepeatedly bytes = case BS.splitAt 32 (hmacSha512 bytes) of
+        (kL, kR) | testBit (kL `BS.index` 31) 5 -> hashRepeatedly (kL <> kR)
+        (kL, kR) -> (kL, kR)
+
+    -- - Clear the lowest 3 bits of the first byte
+    -- - Clear the highest bit of the last byte
+    -- - Set the second highest bit of the last byte
+    --
+    -- As described in:
+    --
+    -- "BIP32-Ed25519 Hierarchical Deterministic Keys over a Non-linear Keyspace".
+    normalizeEd25519 :: ByteString -> ByteString
```

Comment:
It's called buffer "pruning" in [RFC 8032 sec 5.1.5. Key Generation](https://tools.ietf.org/html/rfc8032#section-5.1.5), so it might be better to call it "prune" rather than "normalize".

Reply:
I kept the same terminology as the Rust guys. But I am happy to follow an RFC :)